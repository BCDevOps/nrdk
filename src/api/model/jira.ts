import {RepositoryReference} from '../service/axios-bitbucket-client'

export enum IssueTypeNames {
  RFC = 'RFC',
  RFD = 'RFD',
  RFDSubtask = 'RFD-subtask'
}

export interface IssueReferenceByUrl {
  url: string;
}

export interface ReferenceByKey {
  key: string;
}

export interface PullRequestReference {
  url: string;
  number?: string;
  sourceBranch: string;
  targetBranch: string;
  repository: RepositoryReference;
}

export interface NameAndId {
  id?: string;
  name?: string;
}
export interface ValueAndId {
  id?: string;
  value?: string;
}

export type IssueResolution = NameAndId

export type IssueStatus = NameAndId

export type IssueType = NameAndId

export type Version = NameAndId

export interface Issue {
  id?: string;
  key?: string;
  fields?: {
    project?: ReferenceByKey;
    summary?: string;
    description?: string;
    status?: IssueStatus;
    issuetype?: IssueType;
    fixVersions?: Version[];
    /** Target Environment */
    customfield_10121?: ValueAndId;
    labels?: string[];
    issuelinks?: any[];
    [key: string]: any;
  };
}
export type IssueReference = Issue

export interface DeploymentArgument {
  issue: IssueReference;
  pullRequest: PullRequestReference;
  targetEnvironment: string | string[];
  dryrun?: boolean;
}
export type StartDeploymentArgument = DeploymentArgument

export interface StartDeploymentResult {
  issues: Issue[];
  errors?: any[];
}
export type ProjectReference = ReferenceByKey
