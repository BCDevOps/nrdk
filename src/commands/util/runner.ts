import {streamOutput} from '../../util/child-process'
import {SpawnOptions} from 'child_process'
import {Terraform} from '../../tools/terraform'

export async function runner(argv: string[], spawnOptions?: SpawnOptions) {
  const op: SpawnOptions = spawnOptions || {stdio: ['ignore',  'pipe', 'pipe']}
  const tf = new Terraform()
  return tf.run(argv, op)
  .then(streamOutput(process.stdout, process.stderr))
  .then(proc => proc.status as number)
}
