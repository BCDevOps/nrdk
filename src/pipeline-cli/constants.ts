'use strict'

export class CONSTANTS {
  KINDS: any = {
    LIST: 'List',
    BUILD: 'Build',
    BUILD_CONFIG: 'BuildConfig',
    IMAGE_STREAM: 'ImageStream',
    IMAGE_STREAM_TAG: 'ImageStreamTag',
    IMAGE_STREAM_IMAGE: 'ImageStreamImage',
    DEPLOYMENT_CONFIG: 'DeploymentConfig',
  };

  ENV: any = {
    BUILD_HASH: '_BUILD_HASH',
  };

  LABELS: any = {
    TEMPLATE_HASH: 'template-hash',
    SOURCE_HASH: 'source-hash',
  };

  ANNOTATIONS: any = {
    TEMPLATE_HASH: 'template-hash',
    SOURCE_HASH: 'source-hash',
  };

  POD_PHASES: any = {
    PENDING: 'Pending',
    RUNNING: 'Running',
    SUCCEEDED: 'Succeeded',
    FAILED: 'Failed',
    UNKNOWN: 'Unknown',
  }
}
