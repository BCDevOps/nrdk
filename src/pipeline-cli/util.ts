
/* eslint-disable no-inner-declarations */
/* eslint-disable no-useless-escape */
/* eslint-disable complexity */

import {spawnSync, SpawnSyncOptionsWithStringEncoding, SpawnSyncReturns} from 'child_process'
import * as crypto from 'crypto'
import * as fs from 'fs'
import * as path from 'path'

export namespace Util {
  export function isArray(value: any): boolean {
    const {isArray} = Array
    return isArray(value)
  }

  export function isPlainObject(value: any): boolean {
    return require('lodash.isplainobject')(value)
  }

  export function isString(value: any): boolean {
    return require('lodash.isstring')(value)
  }

  export function normalizeKind(kind: any): any {
    if (kind === 'ImageStream') {
      return 'imagestream.image.openshift.io'
    }
    if (kind === 'BuildConfig') {
      return 'buildconfig.build.openshift.io'
    }
    return kind
  }

  export function isUrl(string: string): boolean {
    const isString = require('lodash.isstring')
    const protocolAndDomainRE = /^(?:\w+)+:\/\/(\S+)$/
    if (!isString(string)) return false
    const match = string.match(protocolAndDomainRE)
    if (!match) return false
    return true
  }

  export function name(resource: any): string {
    if (resource.kind && resource.name) return `${normalizeKind(resource.kind)}/${resource.name}`
    return `${normalizeKind(resource.kind)}/${resource.metadata.name}`
  }

  export function parseName(name: string, defaultNamespace?: string) {
    const nameRegexPattern = '^(?:([^/]+?)/)?(([^/]+?)/(.*?))$'
    const result: any = new RegExp(nameRegexPattern, 'g').exec(name)
    return {
      namespace: result[1] || defaultNamespace,
      kind: result[3],
      name: result[4],
    }
  }

  function unsafeExecSync(command: string, args?: ReadonlyArray<string>, options?: SpawnSyncOptionsWithStringEncoding): SpawnSyncReturns<string> {
    const ret = spawnSync(command, args, options)
    return ret as unknown as SpawnSyncReturns<string>
  }
  /*
  function execSync(command: string, args?: readonly string[], options?: SpawnSyncOptionsWithStringEncoding) {
    const ret = unsafeExecSync(command, args, options)
    if (ret.status !== 0) {
      throw new Error(
        `Failed running '${command} ${args?.join(' ')}' as it returned ${ret.status}`,
      )
    }
    return ret
  }
  */

  export function hashString(itemAsString: any): string {
    const shasum = crypto.createHash('sha1')
    shasum.update(`blob ${itemAsString.length + 1}\0${itemAsString}\n`)

    return shasum.digest('hex')
  }

  export function hashObject(resource: any): string {
    const itemAsString = JSON.stringify(resource)
    return hashString(itemAsString)
  }

  export function fullName(resource: any): string {
    if (resource.namespace && resource.kind && resource.name) {
      return `${resource.namespace}/${normalizeKind(resource.kind)}/${resource.name}`
    }
    return `${resource.metadata.namespace}/${normalizeKind(resource.kind)}/${resource.metadata.name}`
  }

  export function execSync(command: string, args: string[], options: any) {
    const ret = unsafeExecSync(command, args, options)
    if (ret.status !== 0) {
      throw new Error(
        `Failed running '${command} ${args.join(' ')}' (${options.cwd}) as it returned ${ret.status}`,
      )
    }
    return ret
  }

  function _hashDirectory(dir: string) {
    const result: string[] = []
    const items: string[] = fs.readdirSync(dir).sort()

    items.forEach(item => {
      const fullpath = path.join(dir, item)
      const stat = fs.statSync(fullpath)
      if (stat.isDirectory()) {
        result.push(..._hashDirectory(fullpath))
      } else {
        result.push(hashString(fs.readFileSync(fullpath)))
      }
    })
    return result
  }

  export function hashDirectory(dir: string): string {
    const items: string[] = _hashDirectory(dir)
    return hashObject(items)
  }

  function getBuildConfigStrategy(bc: any): any {
    return bc.spec.strategy.sourceStrategy || bc.spec.strategy.dockerStrategy
  }

  export function getBuildConfigInputImages(bc: any): any[] {
    const result = []
    const buildStrategy = getBuildConfigStrategy(bc)

    if (buildStrategy.from) {
      result.push(buildStrategy.from)
    }

    if ((bc.spec.source || {}).images) {
      const sourceImages: any[] = bc.spec.source.images
      sourceImages.forEach(sourceImage => {
        result.push(sourceImage.from)
      })
    }

    return result
  }

  export function applyArgumentsDefaults(options: any) {
    options.git = options.git || {}
    const git = options.git

    if (!git.dir) {
      const gitCmd = unsafeExecSync('git', ['rev-parse', '--show-toplevel'], {encoding: 'utf-8'})
      if (gitCmd.status === 0) {
        git.dir = gitCmd.stdout.trim()
      }
    }

    if (!options.cwd) {
      options.cwd = git.dir
    }

    git.branch = git.branch || {}

    if (!git.branch.name) {
      const gitCmd = unsafeExecSync('git', ['rev-parse', '--abbrev-ref', 'HEAD'], {encoding: 'utf-8'})
      if (gitCmd.status === 0) {
        git.branch.name = gitCmd.stdout.trim()
      }
    }

    if (!git.branch.remote) {
      const gitConfigBranchRemote = unsafeExecSync('git', ['config', `branch.${git.branch.name}.remote`], {encoding: 'utf-8'})
      if (gitConfigBranchRemote.status === 0) {
        git.branch.remote = gitConfigBranchRemote.stdout.trim()
      } else {
      // Default to "origin"
        git.branch.remote = 'origin'
      }
    }

    if (!git.url) {
      const gitCmd = unsafeExecSync('git', ['config', '--get', `remote.${git.branch.remote}.url`], {encoding: 'utf-8'})
      if (gitCmd.status === 0) {
        git.url = gitCmd.stdout.trim()
      }
    }

    git.uri = git.url
    if (!git.http_url && git.url?.length > 0) {
      git.http_url = git.url.replace(
        /((https:\/\/github\.com\/)|(git@github.com:))([^/]+)\/(.*)/,
        'https://github.com/$4/$5',
      )
    }

    if (git.http_url?.startsWith('https://github.com') && !git.branch?.merge) {
      git.branch.merge = `refs/pull/${git.pull_request}/head`
    }

    if (!git.branch.merge) {
      const gitCmd = unsafeExecSync('git', ['config', `branch.${git.branch.name}.merge`], {encoding: 'utf-8'})
      if (gitCmd.status === 0) {
        git.branch.merge = gitCmd.stdout.trim()
      }
    }

    if (!git.owner && git.url) {
      git.owner = git.url.replace(
        /((https:\/\/github\.com\/)|(git@github.com:))([^/]+)\/(.*)/,
        '$4',
      )
    }
    if (!git.repository && git.url) {
      git.repository = git.url.replace(
        /((https:\/\/github\.com\/)|(git@github.com:))([^/]+)\/([^\.]+)\.git/,
        '$5',
      )
    }

    if (options.pr) {
      git.pull_request = options.pr
    }
    // when --ref flag is used
    if (options.ref) {
      git.ref = options.ref
    }

    if (!git.ref) {
      if (git.pull_request) {
        git.ref = `refs/pull/${git.pull_request}/head`
      } else if (git.branch.merge) {
        git.ref = git.branch.merge
      }
    }
    if (!git.ref) {
      git.branch_ref = git.ref
    }
    return options
  }

  export function parseArgumentsFromArray(...argv: string[]) {
    const git = {}
    const options: any = {git}

    argv.forEach(value => {
      if (value.startsWith('--')) {
        value = value.substr(2)
        const sep = value.indexOf('=')
        const argName = value.substring(0, sep)
        const argValue = value.substring(sep + 1)
        if (argName.startsWith('git.')) {
          const ctxPath = argName.substr(4).split('.')
          let ctx: any = git
          ctxPath.forEach((key, index) => {
            if (index === ctxPath.length - 1) {
              ctx[key] = argValue
            } else {
              ctx[key] = ctx[key] || {}
              ctx = ctx[key]
            }
          })
        } else {
          options[argName] = argValue
        }
      }
    })
    return applyArgumentsDefaults(options)
  }

  export function normalizeName(name: string) {
    if (name.startsWith('ImageStream/')) {
      return `imagestream.image.openshift.io/${name.substr('ImageStream/'.length)}`
    }
    if (name.startsWith('BuildConfig/')) {
      return `buildconfig.build.openshift.io/${name.substr('BuildConfig/'.length)}`
    }
    return name
  }

  export function parseArguments() {
    return parseArgumentsFromArray(...process.argv.slice(2))
  }
}

