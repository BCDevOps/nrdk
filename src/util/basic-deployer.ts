/* eslint-disable valid-jsdoc */
import {FlagNames} from '../flags'

const {OpenShiftClientX} = require('@bcgov/pipeline-cli')

export class BasicDeployer {
  settings: any

  constructor(settings: any) {
    this.settings = settings
  }

  /**
    *  @returns any[] an array on openshift resources
  **/
  processTemplates(_oc: any): any[] {
    return []
  }

  async deploy() {
    if (this.settings.options[FlagNames.RFC_VALIDATION] === true) {
      //
    }
    await this.deployOpenshift()
  }

  async deployOpenshift() {
    const settings = this.settings
    const phases = settings.phases
    const options = settings.options
    const phase = settings.options.env
    const changeId = phases[phase].changeId
    const oc = new OpenShiftClientX(Object.assign({namespace: phases[phase].namespace}, options))
    const objects = this.processTemplates(oc)

    oc.applyRecommendedLabels(objects, phases[phase].name, phase, `${changeId}`, phases[phase].instance)
    oc.importImageStreams(objects, phases[phase].tag, phases.build.namespace, phases.build.tag)
    return oc.applyAndDeploy(objects, phases[phase].instance)
  }
}
