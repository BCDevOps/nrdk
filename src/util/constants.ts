'use strict'
export const ISSUE_TYPE_CODE = Object.freeze({RFD: 'RFD', RFD_SUBTASK: 'RFD-subtask'})

export const ISSUE_LINK_TYPE_NAME = Object.freeze({RFC_FRD: 'RFC-RFD'})

export const ENV = Object.freeze({BUILD: 'build', DEV: 'dev', DLVR: 'dlvr', TEST: 'test', PROD: 'prod'})
export const JIRA_TARGET_ENV = Object.freeze({DLVR: 'DLVR', TEST: 'TEST', PROD: 'PROD'})

export const JIRA_ENV_REVIEWER = Object.freeze([
  Object.freeze({ENV: JIRA_TARGET_ENV.DLVR, REVIEWER: 'Developer'}),
  Object.freeze({ENV: JIRA_TARGET_ENV.TEST, REVIEWER: 'IIT'}),
  Object.freeze({ENV: JIRA_TARGET_ENV.PROD, REVIEWER: 'Business'}),
])

export const DB_ACTION = Object.freeze({BACKUP: 'backup', RECOVERY: 'recovery'})

export const VERIFY_STATUS = Object.freeze({READY: 'Ready', NOT_READY: 'Not Ready'})
export const REASON = Object.freeze({
  REASON_CODE_RFD_BLOCKED: 'RFD_BLOCKED',
  REASON_DESC_RFD_BLOCKED: 'RFD(s) are blocked',
  REASON_CODE_RFD_NOT_APPROVED: 'RFD_NOT_APPROVED',
  REASON_DESC_RFD_NOT_APPROVED: 'RFD(s) are not approved',
  REASON_CODE_PREVIOUS_RFD_NOT_CLOSED: 'PREVIOUS_RFD_NOT_CLOSED',
  REASON_DESC_PREVIOUS_RFD_NOT_CLOSED: 'Previous stage RFD(s) are not closed',
  REASON_CODE_RFC_NOT_AUTHORIZED: 'RFC_NOT_AUTHORIZED',
  REASON_DESC_RFC_NOT_AUTHORIZED: 'RFC is not authorized to target environment',
})
