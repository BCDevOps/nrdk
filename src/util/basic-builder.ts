/* eslint-disable valid-jsdoc */
/* eslint-disable max-params */
/* eslint-disable unicorn/import-index */
import {OpenShiftClientX, GitOperation} from '../'
import {AxiosJiraClient} from '../api/service/axios-jira-client'
import {RfdHelper} from '../util/rfd-helper'
import {AxiosBitBucketClient} from '../api/service/axios-bitbucket-client'
import {FlagNames} from '../flags'
export class BasicBuilder {
    settings: any

    constructor(settings: any) {
      this.settings = settings
    }

    processTemplates(_oc: any): any[] {
      return []
    }

    async build() {
      const settings = this.settings
      const phases = settings.phases
      const env = 'build'
      const username = settings.phases[env].credentials.idir.user
      const password = settings.phases[env].credentials.idir.pass
      Object.assign(settings.options.git, {credentials: {username: username, password: password}})
      const sourceBranch = settings.options.git.branch.merge
      // const EMPTY = ''
      const targetBranch = (settings?.options?.git?.change?.target || '').trim()

      // if (targetBranch && targetBranch.toLowerCase() === 'master') {
      // Build for target = 'master'
      const repo = AxiosBitBucketClient.parseUrl(settings.options.git.url)
      const issueKey =  await AxiosJiraClient.parseJiraIssueKeyFromUri(settings.options.git.branch.merge)

      if (this.settings.options[FlagNames.RFC_VALIDATION] === true) {
        const helper = new RfdHelper({})
        await helper.createDeployments({
          issue: {key: issueKey},
          pullRequest: {
            url: `https://apps.nrs.gov.bc.ca/int/stash/projects/${repo.project.key}/repos/${repo.slug}/pull-requests/${settings.options.pr}/overview`,
            number: settings.options.pr,
            sourceBranch: sourceBranch,
            targetBranch: targetBranch,
            repository: repo,
          },
          targetEnvironment: settings.environments,
        }).then(issues => {
          helper.print(issues)
        })
      }
      // await this._createJiraAutoRFDs(jiraUrl, repoName, changeBranch, branchName, username, password)

      // await this._gitTargetSyncVerify()
      // } else if (targetBranch !== EMPTY) {
      //  // Build for target = 'other branch'
      //  await this._gitTargetSyncVerify()
      // }

      const oc = new OpenShiftClientX(Object.assign({namespace: phases.build.namespace}, settings.options))
      const processedTemplate = this.processTemplates(oc)
      const phase = env
      oc.applyRecommendedLabels(
        processedTemplate,
        phases[phase].name,
        phase,
        phases[phase].changeId,
        phases[phase].instance
      )
      oc.applyAndBuild(processedTemplate)
      // const git = new GitOperation(this.settings.options.git)
      // return git.clear()
    } // end build

    /**
     * Calling Git module to verify if both change/target branches are not out of sync.
     */
    async _gitTargetSyncVerify() {
      const git = new GitOperation(this.settings.options.git)
      if (await git.isTargetBranchOutofSync()) {
        // eslint-disable-next-line no-console
        console.log('Successfully Verified that branch is not out of sync with target')
      }
    }
}
