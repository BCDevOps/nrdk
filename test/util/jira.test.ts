'use strict'
import {Jira} from '../../src/util/jira'
import {createSandbox} from 'sinon'
const sandbox = createSandbox()
import {isActive, activate, disableNetConnect, restore, cleanAll} from 'nock'
import expect, {extend} from 'expect'
import {toMatchSnapshot} from './expect-mocha-snapshot'
extend({toMatchSnapshot})

import {STATUS_RESOLVED} from '../../src/util/jira-rfd-workflow-v1.2.2'
import {STATUS_APPROVED} from '../../src/util/jira-rfc-workflow-v2.0.0'
import {ENV} from '../../src/util/constants'
import {JiraMock} from './jira-mock'
import {Context, Suite} from 'mocha'

const JIRA_SETTINGS = {
  url: 'bwa.nrs.gov.bc.ca/int/jira',
  username: 'fake',
  password: 'fake',
  rfcIssueKey: 'FAKE-123',
  changeBranch: 'FAKE-123-rfc',
  branchName: 'PR-456',
  repoName: 'FAKE',
  projectName: 'FAKE',
  version: '1.0.0',
}

describe('Jira @jira', function () {
  beforeEach('Using fake settings to create JIRA object', function () {
    if (!isActive()) {
      activate()
    }
    disableNetConnect()
  })
  afterEach('Completely restore all fakes created through the sandbox', function () {
    sandbox.restore()
    restore()
    cleanAll()
  })
  describe('RFDs and RFD-Subtasks', function (this: Suite) {
    it('create', async function (this: Context) {
      const jira = new Jira(Object.assign({}, {phase: 'jira-update', jira: JIRA_SETTINGS}))
      const jiraMock = new JiraMock(jira)
      // RFC already exists
      jiraMock._addIssue({
        key: 'FAKE-123',
        fields: {
          fixVersions: ['0.0.0'],
          issuetype: {id: '10400', name: 'RFC'},
        },
      })
      jiraMock.start()

      await expect(jira.createRFD()).resolves.toBe(true)
      expect([...jiraMock.issues.values()]).toHaveLength(7)
      // console.dir(Array.from(issues.values()), { depth: 4 })
      expect([...jiraMock.issues.values()]).toMatchSnapshot(this as any, 'a7c66bfb-a011-4adc-9b10-6dda737a2d5b')
      expect(jiraMock.links).toHaveLength(3)
      // console.dir(links)
      expect(jiraMock.links).toMatchSnapshot(this as any, '72829b90-f2e8-4569-92d8-10a7c889c8cc')
    })
    it.skip('update', async function () {
      const initialState = [
        {
          key: 'FAKE-123',
          fields: {
            fixVersions: ['0.0.0'],
            issuetype: {id: '10400', name: 'RFC'},
            status: STATUS_APPROVED,
            issuelinks: [
              {
                inwardIssue: {key: 'FAKE-123'},
                outwardIssue: {key: 'FAKE-1000'},
                type: {id: '10300'},
              },
              {
                inwardIssue: {key: 'FAKE-123'},
                outwardIssue: {key: 'FAKE-1002'},
                type: {id: '10300'},
              },
              {
                inwardIssue: {key: 'FAKE-123'},
                outwardIssue: {key: 'FAKE-1004'},
                type: {id: '10300'},
              },
            ],
          },
        },
        {
          fields: {
            project: {key: 'FAKE'},
            issuetype: {name: 'RFD'},
            customfield_10121: {value: 'DLVR'},
            labels: ['FAKE', 'auto'],
            fixVersions: [{}],
            description: 'Deploying changes from PR NO: PR-456 in REPO: FAKE',
            summary: 'RFD-DLVR-FAKE-123-rfc-PR-456',
            status: STATUS_RESOLVED,
          },
          key: 'FAKE-1000',
        },
        {
          fields: {
            project: {key: 'FAKE'},
            issuetype: {name: 'RFD-subtask'},
            customfield_10121: {value: 'DLVR'},
            parent: {key: 'FAKE-1000'},
            summary: 'RFD-Subtask-DLVR-FAKE-123-rfc-PR-456-Developer-Review',
            components: [{name: 'FAKE'}],
            status: STATUS_RESOLVED,
          },
          key: 'FAKE-1001',
        },
        {
          fields: {
            project: {key: 'FAKE'},
            issuetype: {name: 'RFD'},
            customfield_10121: {value: 'TEST'},
            labels: ['FAKE', 'auto'],
            fixVersions: [{}],
            description: 'Deploying changes from PR NO: PR-456 in REPO: FAKE',
            summary: 'RFD-TEST-FAKE-123-rfc-PR-456',
            status: STATUS_RESOLVED,
          },
          key: 'FAKE-1002',
        },
        {
          fields: {
            project: {key: 'FAKE'},
            issuetype: {name: 'RFD-subtask'},
            customfield_10121: {value: 'TEST'},
            parent: {key: 'FAKE-1002'},
            summary: 'RFD-Subtask-TEST-FAKE-123-rfc-PR-456-IIT-Review',
            components: [{name: 'FAKE'}],
            status: STATUS_RESOLVED,
          },
          key: 'FAKE-1003',
        },
        {
          fields: {
            project: {key: 'FAKE'},
            issuetype: {name: 'RFD'},
            customfield_10121: {value: 'PROD'},
            labels: ['FAKE', 'auto'],
            fixVersions: [{}],
            description: 'Deploying changes from PR NO: PR-456 in REPO: FAKE',
            summary: 'RFD-PROD-FAKE-123-rfc-PR-456',
            status: STATUS_RESOLVED,
          },
          key: 'FAKE-1004',
        },
        {
          fields: {
            project: {key: 'FAKE'},
            issuetype: {name: 'RFD-subtask'},
            customfield_10121: {value: 'PROD'},
            parent: {key: 'FAKE-1004'},
            summary: 'RFD-Subtask-PROD-FAKE-123-rfc-PR-456-Business-Review',
            components: [{name: 'FAKE'}],
            status: STATUS_RESOLVED,
          },
          key: 'FAKE-1005',
        },
      ]
      const jira = new Jira(Object.assign({}, {phase: 'jira-update', jira: JIRA_SETTINGS}))
      const jiraMock = new JiraMock(jira)
      jiraMock.start()
      jiraMock._addIssue(initialState)
      expect(await jira.createRFD()).toBe(true)
      // console.dir(JSON.stringify(Array.from(jiraMock.issues.values())))
      await expect([...jiraMock.issues.values()]).toMatchSnapshot(
        this as any,
        '3e7704c4-f8d6-44f5-8473-4c7a0c094633'
      )

      await jira.transitionRFDpostDeployment(ENV.DLVR)
      await jira.transitionRFDpostDeployment(ENV.TEST)
      await jira.transitionRFDpostDeployment(ENV.PROD)
    })
  })
})
