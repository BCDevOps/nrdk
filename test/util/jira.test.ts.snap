// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Jira @jira RFDs and RFD-Subtasks create: 72829b90-f2e8-4569-92d8-10a7c889c8cc 1`] = `
Array [
  Object {
    "inwardIssue": Object {
      "key": "FAKE-123",
    },
    "outwardIssue": Object {
      "key": "FAKE-1000",
    },
    "type": Object {
      "id": "10300",
    },
  },
  Object {
    "inwardIssue": Object {
      "key": "FAKE-123",
    },
    "outwardIssue": Object {
      "key": "FAKE-1002",
    },
    "type": Object {
      "id": "10300",
    },
  },
  Object {
    "inwardIssue": Object {
      "key": "FAKE-123",
    },
    "outwardIssue": Object {
      "key": "FAKE-1004",
    },
    "type": Object {
      "id": "10300",
    },
  },
]
`;

exports[`Jira @jira RFDs and RFD-Subtasks create: a7c66bfb-a011-4adc-9b10-6dda737a2d5b 1`] = `
Array [
  Object {
    "fields": Object {
      "fixVersions": Array [
        "0.0.0",
      ],
      "issuelinks": Array [
        Object {
          "inwardIssue": Object {
            "key": "FAKE-123",
          },
          "outwardIssue": Object {
            "key": "FAKE-1000",
          },
          "type": Object {
            "id": "10300",
          },
        },
        Object {
          "inwardIssue": Object {
            "key": "FAKE-123",
          },
          "outwardIssue": Object {
            "key": "FAKE-1002",
          },
          "type": Object {
            "id": "10300",
          },
        },
        Object {
          "inwardIssue": Object {
            "key": "FAKE-123",
          },
          "outwardIssue": Object {
            "key": "FAKE-1004",
          },
          "type": Object {
            "id": "10300",
          },
        },
      ],
      "issuetype": Object {
        "id": "10400",
        "name": "RFC",
      },
      "subtasks": Array [],
    },
    "key": "FAKE-123",
  },
  Object {
    "fields": Object {
      "customfield_10121": Object {
        "value": "DLVR",
      },
      "description": "Deploying changes from PR NO: PR-456 in REPO: FAKE",
      "fixVersions": Array [
        Object {},
      ],
      "issuelinks": Array [
        Object {
          "inwardIssue": Object {
            "key": "FAKE-123",
          },
          "outwardIssue": Object {
            "key": "FAKE-1000",
          },
          "type": Object {
            "id": "10300",
          },
        },
      ],
      "issuetype": Object {
        "name": "RFD",
      },
      "labels": Array [
        "FAKE",
        "auto",
      ],
      "project": Object {
        "key": "FAKE",
      },
      "status": Object {
        "id": "10316",
        "name": "Submitted",
      },
      "subtasks": Array [],
      "summary": "RFD-DLVR-FAKE-123-rfc-PR-456",
    },
    "key": "FAKE-1000",
  },
  Object {
    "fields": Object {
      "components": Array [
        Object {
          "name": "FAKE",
        },
      ],
      "customfield_10121": Object {
        "value": "DLVR",
      },
      "issuelinks": Array [],
      "issuetype": Object {
        "name": "RFD-subtask",
      },
      "parent": Object {
        "key": "FAKE-1000",
      },
      "project": Object {
        "key": "FAKE",
      },
      "status": Object {
        "id": "10316",
        "name": "Submitted",
      },
      "subtasks": Array [],
      "summary": "RFD-Subtask-DLVR-FAKE-123-rfc-PR-456-Developer-Review",
    },
    "key": "FAKE-1001",
  },
  Object {
    "fields": Object {
      "customfield_10121": Object {
        "value": "TEST",
      },
      "description": "Deploying changes from PR NO: PR-456 in REPO: FAKE",
      "fixVersions": Array [
        Object {},
      ],
      "issuelinks": Array [
        Object {
          "inwardIssue": Object {
            "key": "FAKE-123",
          },
          "outwardIssue": Object {
            "key": "FAKE-1002",
          },
          "type": Object {
            "id": "10300",
          },
        },
      ],
      "issuetype": Object {
        "name": "RFD",
      },
      "labels": Array [
        "FAKE",
        "auto",
      ],
      "project": Object {
        "key": "FAKE",
      },
      "status": Object {
        "id": "10316",
        "name": "Submitted",
      },
      "subtasks": Array [],
      "summary": "RFD-TEST-FAKE-123-rfc-PR-456",
    },
    "key": "FAKE-1002",
  },
  Object {
    "fields": Object {
      "components": Array [
        Object {
          "name": "FAKE",
        },
      ],
      "customfield_10121": Object {
        "value": "TEST",
      },
      "issuelinks": Array [],
      "issuetype": Object {
        "name": "RFD-subtask",
      },
      "parent": Object {
        "key": "FAKE-1002",
      },
      "project": Object {
        "key": "FAKE",
      },
      "status": Object {
        "id": "10316",
        "name": "Submitted",
      },
      "subtasks": Array [],
      "summary": "RFD-Subtask-TEST-FAKE-123-rfc-PR-456-IIT-Review",
    },
    "key": "FAKE-1003",
  },
  Object {
    "fields": Object {
      "customfield_10121": Object {
        "value": "PROD",
      },
      "description": "Deploying changes from PR NO: PR-456 in REPO: FAKE",
      "fixVersions": Array [
        Object {},
      ],
      "issuelinks": Array [
        Object {
          "inwardIssue": Object {
            "key": "FAKE-123",
          },
          "outwardIssue": Object {
            "key": "FAKE-1004",
          },
          "type": Object {
            "id": "10300",
          },
        },
      ],
      "issuetype": Object {
        "name": "RFD",
      },
      "labels": Array [
        "FAKE",
        "auto",
      ],
      "project": Object {
        "key": "FAKE",
      },
      "status": Object {
        "id": "10316",
        "name": "Submitted",
      },
      "subtasks": Array [],
      "summary": "RFD-PROD-FAKE-123-rfc-PR-456",
    },
    "key": "FAKE-1004",
  },
  Object {
    "fields": Object {
      "components": Array [
        Object {
          "name": "FAKE",
        },
      ],
      "customfield_10121": Object {
        "value": "PROD",
      },
      "issuelinks": Array [],
      "issuetype": Object {
        "name": "RFD-subtask",
      },
      "parent": Object {
        "key": "FAKE-1004",
      },
      "project": Object {
        "key": "FAKE",
      },
      "status": Object {
        "id": "10316",
        "name": "Submitted",
      },
      "subtasks": Array [],
      "summary": "RFD-Subtask-PROD-FAKE-123-rfc-PR-456-Business-Review",
    },
    "key": "FAKE-1005",
  },
]
`;
